lib/ts/recipe/totp/types.ts

```diff
@@ -0,0 +1,143 @@
+
+export type TypeInput = {
+    issuer: string;
+    defaultSkew?: number;
+    defaultPeriod?: number;
+    allowUnverifiedDevice?: boolean;
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    issuer: string;
+    defaultSkew: number;
+    defaultPeriod: number;
+    allowUnverifiedDevices: boolean;
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type RecipeInterface = {
+    createDevice: (input: {
+        userId: string;
+        deviceName: string;
+        skew?: number;
+        period?: number;
+        userContext: any;
+    }) => Promise<{ status: "OK"; qr: string } | { status: "DEVICE_ALREADY_EXISTS_ERROR" }>;
+    verifyCode: (input: {
+        userId: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" | "TOTP_NOT_ENABLED_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    verifyDevice: (input: {
+        userId: string;
+        deviceName: string;
+        totp: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK" | "INVALID_TOTP_ERROR" | "TOTP_NOT_ENABLED_ERROR" | "UNKNOWN_DEVICE_ERROR" }
+        | { status: "LIMIT_REACHED_ERROR"; retryAfterMs: number }
+    >;
+    updateDevice: (input: {
+        userId: string;
+        existingDeviceName: string;
+        newDeviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "TOTP_NOT_ENABLED_ERROR" | "DEVICE_ALREADY_EXISTS_ERROR" | "UNKNOWN_DEVICE_ERROR" }>;
+    removeDevice: (input: {
+        userId: string;
+        deviceName: string;
+        userContext: any;
+    }) => Promise<{ status: "OK" | "TOTP_NOT_ENABLED_ERROR" }>;
+    listDevices: (input: {
+        userId: string;
+        userContext: any;
+    }) => Promise<
+        | { status: "OK"; devices: { name: string; period: number; skew: number; verified: boolean }[] }
+        | { status: "TOTP_NOT_ENABLED_ERROR" }
+    >;
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    appInfo: NormalisedAppinfo;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    createDevicePOST?: (input: {
+        session: SessionContainer;
+        deviceName: string;
+        period: number;
+        skew: number;
```

Comment:
this does not come from the frontend right? So we do not need these as part of the input to this API?

Reply:
Can come from the frontend. But it can complicate things for the end user. So we can keep it optional?

Reply:
I see no purpose on having this come from the frontend at all. Remove this

Reply:
Removed.

Reply:
You said this is removed, but it's still there in the API input, but just made optional. Remove it entirely from the API input?

Reply:
Removed from API input.